Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -109,49 +114,129 @@ private def normEDS'' (b c d : R) : ℕ → R
+  simp only [m.mul_div_cancel_left two_pos]
+
+lemma preNormEDS'_even (m : ℕ) : preNormEDS' b c d (2 * (m + 3)) =
+    preNormEDS' b c d (m + 2) ^ 2 * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 5) -
+    preNormEDS' b c d (m + 1) * preNormEDS' b c d (m + 3) * preNormEDS' b c d (m + 4) ^ 2 := by
+  rw [show 2 * (m + 3) = 2 * m + 1 + 5 by rfl, preNormEDS', dif_neg m.not_even_two_mul_add_one]
+  simp only [Nat.mul_add_div two_pos]
+  rfl
+
+/-- The auxiliary sequence for a normalised EDS `W : ℤ → R`,
+with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = 1`, `W(3) = c`, and `W(4) = d`.
+
+This extends `preNormEDS'` by defining its values at negative integers. -/
+def preNormEDS (n : ℤ) : R :=
+  n.sign * preNormEDS' b c d n.natAbs
+
+@[simp]
+lemma preNormEDS_ofNat (n : ℕ) : preNormEDS b c d n = preNormEDS' b c d n := by
+  by_cases hn : n = 0
+  · rw [hn, preNormEDS, Nat.cast_zero, Int.sign_zero, Int.cast_zero, zero_mul, preNormEDS'_zero]
+  · rw [preNormEDS, Int.sign_coe_nat_of_nonzero hn, Int.cast_one, one_mul, Int.natAbs_cast]
+
+@[simp]
+lemma preNormEDS_zero : preNormEDS b c d 0 = 0 := by
+  erw [preNormEDS_ofNat, preNormEDS'_zero]
+
+@[simp]
+lemma preNormEDS_one : preNormEDS b c d 1 = 1 := by
+  erw [preNormEDS_ofNat, preNormEDS'_one]
+
+@[simp]
+lemma preNormEDS_two : preNormEDS b c d 2 = 1 := by
+  erw [preNormEDS_ofNat, preNormEDS'_two]
+
+@[simp]
+lemma preNormEDS_three : preNormEDS b c d 3 = c := by
+  erw [preNormEDS_ofNat, preNormEDS'_three]
+
+@[simp]
+lemma preNormEDS_four : preNormEDS b c d 4 = d := by
+  erw [preNormEDS_ofNat, preNormEDS'_four]
+
+lemma preNormEDS_odd (m : ℕ) : preNormEDS b c d (2 * (m + 2) + 1) =
+    preNormEDS b c d (m + 4) * preNormEDS b c d (m + 2) ^ 3 * (if Even m then b else 1) -
+    preNormEDS b c d (m + 1) * preNormEDS b c d (m + 3) ^ 3 * (if Even m then 1 else b) := by
+  repeat erw [preNormEDS_ofNat]
+  exact preNormEDS'_odd ..
+
+lemma preNormEDS_even (m : ℕ) : preNormEDS b c d (2 * (m + 3)) =
+    preNormEDS b c d (m + 2) ^ 2 * preNormEDS b c d (m + 3) * preNormEDS b c d (m + 5) -
+    preNormEDS b c d (m + 1) * preNormEDS b c d (m + 3) * preNormEDS b c d (m + 4) ^ 2 := by
+  repeat erw [preNormEDS_ofNat]
+  exact preNormEDS'_even ..
+
+@[simp]
+lemma preNormEDS_neg (n : ℤ) : preNormEDS b c d (-n) = -preNormEDS b c d n := by
+  rw [preNormEDS, Int.sign_neg, Int.cast_neg, neg_mul, Int.natAbs_neg, preNormEDS]
+
 /-- The canonical example of a normalised EDS `W : ℕ → R`,
-with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = b * d`.
+with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = d * b`.
 
-This is defined in terms of a truncated sequence whose even terms differ by a factor of `b`. -/
+This is defined in terms of `preNormEDS'` whose even terms differ by a factor of `b`. -/
 def normEDS' (n : ℕ) : R :=
-  normEDS'' b c d n * if Even n then b else 1
+  preNormEDS' (b ^ 4) c d n * if Even n then b else 1
```

Comment:
Why does `b` get replaced by `b^4` here? The definition of `PreNormEDS'` seems to be the same as that of `normEDS''`.

Reply:
`b^4` are changed to `b` [earlier in the file](https://github.com/leanprover-community/mathlib4/pull/10814/files#diff-02d4668be95cd41b094c2ec5859acf1cf92c74833b9181a533bcce3a23c00961L112-R118).